### PR TITLE
Create basic draft trip reports page #38

### LINKED ISSUE #38: [Frontend] Create basic trip reports page
Create basic trip reports page

---

### 🎯 **Acceptance Criteria**

- [ ] Create src/app/(frontend)/trip-reports/page.tsx.
- [ ] Fetch all published trip reports, sorted by tripDate or creation date (descending).
- [ ] For each report, display: Title, Author (name), Trip Date.
- [ ] Make each title a link to /trip-reports/[slug].
- [ ] Implement pagination.
- [ ] Loading/error states.

---

### 📊 **Definition of Done (DoD)**

- [ ] Acceptance criteria met
- [ ] Agreed tests written
- [ ] Agreed documentation in GitHub wiki
- [ ] PR reviewed

---

### 🔗 **Dependencies**

#33 

---

### PR DESCRIPTION
**Pull request by neryadom**
closes #38

## 🌟 Context 

This pull request introduces a new feature for rendering and displaying trip reports, including a rich text rendering system, a trip reports grid component, pagination, and corresponding tests. The changes are grouped into three themes: rich text rendering, trip reports feature, and testing.

---

### Rich Text Rendering:
* Added a utility function `internalDocToHref` to generate URLs for internal links based on the document's relation type and slug. (`src/app/(frontend)/_components/RichText/converters/internalLink.tsx`)
* Introduced `RichTextRenderer` component to render rich text content using custom converters, including support for internal links. (`src/app/(frontend)/_components/RichText/index.tsx`)

### Trip Reports Feature:
* Created `TripReportCard` component to display individual trip reports, including title, date, location, authors, and content preview using the `RichTextRenderer`. (`src/app/(frontend)/trip-reports/_components/TripReportCard.tsx`)
* Added `TripReportsGrid` component to display a grid of trip reports and `TripReportsGridFallback` for a loading skeleton. (`src/app/(frontend)/trip-reports/_components/TripReportsGrid.tsx`, `src/app/(frontend)/trip-reports/_components/TripReportsGridFallback.tsx`) 
* Updated the `TripReportsPage` to include pagination, a loading state, and integration with the `TripReportsGrid`. (`src/app/(frontend)/trip-reports/page.tsx`, `src/app/(frontend)/trip-reports/loading.tsx`)

### Testing:
* Added unit tests for the `getTripReports` query function to verify default and custom parameters, as well as handling of empty results. (`src/queries/__tests__/tripReports.test.ts`)

### Backend Query:
* Implemented `getTripReports` function to fetch trip reports from the backend with support for pagination and sorting. (`src/queries/tripReports.tsx`)

---

## 📋 Notes 

- Edge cases may involve RichText content breaking due to data types not written in here yet.

- Feedback on implementation of the RichText functionality would be appreciated.

- Feedback on implementation of the pagination functionality would be appreciated.

### CODE REVIEW

File: src/queries/tripReports.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * @param page - The page number to get
+ * @param limit - The number of rivers to get per page
+ * @param sort - The field to sort the rivers by
+ * @returns The trip reports and pagination information
+ */
+export async function getTripReports(
+  page = 1,
+  limit = 10,
+  sort = '-createdAt',
+) {
```
Comment (dyzhuu) [inline src/queries/tripReports.tsx:14]: could you use an object argument here for consistency?

File: src/app/(frontend)/trip-reports/page.tsx [hunk 2]
```diff
@@ -1,12 +1,92 @@
...
+
+export default async function TripReportsPage({
+  searchParams: SearchParamsPromise,
+}: {
+  searchParams: Promise<{ page?: string; limit?: string }>
+}) {
+  const { page: pageStr = '1', limit: limitStr = '8' } =
+    await SearchParamsPromise
+  const page = parseInt(pageStr, 10)
+  const limit = parseInt(limitStr, 10)
```
Comment (dyzhuu) [inline src/app/(frontend)/trip-reports/page.tsx:16]: this doesn't handle NaN values well. for parsing search params I would suggest using zod, which can also handle type coercion, setting default values, catching errors.

(have a look at z.coerce, z.default, z.catch)
https://zod.dev/?id=coercion-for-primitives

and then you can call something like
```
const { page, limit } = searchParamsSchema.parse(await searchParams)
```

the searchParams type can also be inferred from this schema

File: src/app/(frontend)/trip-reports/page.tsx [hunk 3]
```diff
@@ -1,12 +1,92 @@
-export default async function TripReportsPage() {
+import React, { Suspense } from 'react'
+import Link from 'next/link'
+
+import { getTripReports } from '@/queries/tripReports'
+import { TripReportsGrid } from './_components/TripReportsGrid'
+import { TripReportsGridFallback } from './_components/TripReportsGridFallback'
+
+export default async function TripReportsPage({
+  searchParams: SearchParamsPromise,
```
Comment (dyzhuu) [inline src/app/(frontend)/trip-reports/page.tsx:9]: I think the renaming here is unnecessary, also should use camelCase for variable names
```suggestion
  searchParams,
```

File: src/app/(frontend)/trip-reports/page.tsx [hunk 4]
```diff
@@ -1,12 +1,92 @@
...
+    pageLinks.push(1)
+    if (left > 2) pageLinks.push('…')
+  }
+  for (let p = left; p <= right; p++) {
+    pageLinks.push(p)
+  }
+  if (right < totalPages) {
+    if (right < totalPages - 1) pageLinks.push('…')
+    pageLinks.push(totalPages)
+  }
```
Comment (dyzhuu) [inline src/app/(frontend)/trip-reports/page.tsx:41]: This is really good work! But after having a think about it, maybe we'll skip implementing pagination for this ticket. I'd like to make this more robust with a custom pagination hook / component, and will probably make a separate ticket for it -- sorry for the change in scope

File: src/app/(frontend)/_components/RichText/index.tsx
[the remark this answers is not in the document]
Comment (dyzhuu) [reply src/app/(frontend)/_components/RichText/index.tsx:1]: also for components in the shared components folder, use barrel exports

**Review by dyzhuu — CHANGES_REQUESTED** [review]
really amazing stuff! few comments

File: src/app/(frontend)/trip-reports/page.tsx (hunk 2 above)
Comment (dyzhuu) [reply src/app/(frontend)/trip-reports/page.tsx:16]: ^ perhaps you could also write some tests for this functionality?

File: src/app/(frontend)/_components/RichText/index.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+} from '@payloadcms/richtext-lexical/react'
+
+import { internalDocToHref } from './converters/internalLink'
+
+const converters: JSXConvertersFunction = ({ defaultConverters }) => ({
+  ...defaultConverters,
+  ...LinkJSXConverter({ internalDocToHref }),
+})
+
+export function RichTextRenderer({ data }: { data: SerializedEditorState }) {
```
Comment (dyzhuu) [inline src/app/(frontend)/_components/RichText/index.tsx:19]: In this guide (which I assume was the one you followed) they also allow passing in a custom className, which is always nice to have.
https://payloadcms.com/posts/guides/how-to-render-rich-text-from-payload-in-a-nextjs-frontend

File: src/app/(frontend)/_components/RichText/index.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
+// src/components/RichTextRenderer.tsx
+'use client'
```
[the remark this answers is not in the document]
Comment (dyzhuu) [reply src/app/(frontend)/_components/RichText/index.tsx:2]: if this could be made into a server component, you could consider adding a Suspend and ErrorBoundary for loading and error fallbacks.

File: src/app/(frontend)/_components/RichText/index.tsx [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+const converters: JSXConvertersFunction = ({ defaultConverters }) => ({
+  ...defaultConverters,
+  ...LinkJSXConverter({ internalDocToHref }),
+})
+
+export function RichTextRenderer({ data }: { data: SerializedEditorState }) {
+  useEffect(() => {
+    console.log('Rendering rich text:', data)
+  }, [data])
```
[the remark this answers is not in the document]
Comment (neryadom) [reply src/app/(frontend)/_components/RichText/index.tsx:22]: Logging removed

File: src/app/(frontend)/trip-reports/page.tsx (hunk 4 above)
Comment (neryadom) [reply src/app/(frontend)/trip-reports/page.tsx:41]: No worries! Removed

File: src/app/(frontend)/trip-reports/page.tsx (hunk 3 above)
Comment (neryadom) [reply src/app/(frontend)/trip-reports/page.tsx:9]: Noted!

File: src/app/(frontend)/trip-reports/page.tsx (hunk 2 above)
Comment (neryadom) [reply src/app/(frontend)/trip-reports/page.tsx:16]: Removed as per following comment.

File: src/queries/tripReports.tsx (hunk 1 above)
Comment (neryadom) [reply src/queries/tripReports.tsx:14]: Done.

File: src/app/(frontend)/_components/RichText/converters/internalLink.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import { SerializedLinkNode } from '@payloadcms/richtext-lexical'
+
+export const internalDocToHref = ({
```
[the remark this answers is not in the document]
Comment (neryadom) [reply src/app/(frontend)/_components/RichText/converters/internalLink.tsx:3]: Chose to remove that file as I'm not too sure about relevance of internalLink converters. If we were to implement those, surely we would also want to implement converters for media etc... ? Once more, not too sure if what I wrote here was super relevant. Just going to go with the simplest RichTextRenderer in src/components/RichText

File: src/app/(frontend)/_components/RichText/converters/internalLink.tsx [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    // fallback: raw ID or string
+    slug = String(value)
+  }
+
+  // route based on the collection
+  if (relationTo === 'posts') {
+    return `/posts/${slug}`
+  } else if (relationTo === 'users') {
+    return `/users/${slug}`
+  } else {
```
[the remark this answers is not in the document]
Comment (neryadom) [reply src/app/(frontend)/_components/RichText/converters/internalLink.tsx:35]: Not as it is, haha. Removed

File: src/app/(frontend)/_components/RichText/index.tsx
[the remark this answers is not in the document]
Comment (neryadom) [reply src/app/(frontend)/_components/RichText/index.tsx:1]: understood! created a RichText folder in src/components and put the index file in there!

File: src/app/(frontend)/_components/RichText/index.tsx (hunk 5 above)
Comment (neryadom) [reply src/app/(frontend)/_components/RichText/index.tsx:19]: yes, that is the one i looked at!

File: src/app/(frontend)/_components/RichText/converters/internalLink.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (dyzhuu) [reply src/app/(frontend)/_components/RichText/converters/internalLink.tsx]: Makes sense!

File: src/components/index.ts
Comment (dyzhuu) [inline src/components/index.ts:1]: oops i may be using the barrel export term a bit incorrectly, but i meant something like
```sh
/components
  /RichText
    RichText.tsx
    # ... any other related components
    index.ts       # <-- barrel file: `export * from './RichText'`
```

**Review by phawatboom — COMMENTED** [review]
Really amazing stuff!

File: src/app/(frontend)/trip-reports/_components/TripReportCard.tsx [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import React, { Suspense } from 'react'
+import Image from 'next/image'
+import Link from 'next/link'
+
+import { TripReport } from '@/payload-types'
+import type { Exec } from '@/payload-types'
+import { RichTextRenderer } from '../../../../components/RichText/index'
```
[the remark this answers is not in the document]
Comment (neryadom) [reply src/app/(frontend)/trip-reports/_components/TripReportCard.tsx:7]: Done!

File: src/components/index.ts
[the remark this answers is not in the document]
Comment (neryadom) [reply src/components/index.ts:1]: removed `src/component/index.ts` and remade it in `src/component/RichText`

File: src/app/(frontend)/trip-reports/_components/TripReportCard.tsx [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      )}
+      {postAuthors.length > 0 && (
+        <p className="mt-1 text-sm text-gray-500">
+          By {postAuthors.map((a) => a.name).join(', ')}
+        </p>
+      )}
+
+      {report.content && (
+        <div className="mt-1 line-clamp-2 text-sm whitespace-pre-wrap text-gray-600">
+          <Suspense fallback={<p>Loading content…</p>}>
```
[the remark this answers is not in the document]
Comment (neryadom) [reply src/app/(frontend)/trip-reports/_components/TripReportCard.tsx:56]: No worries! Readjusted.

File: src/app/(frontend)/trip-reports/_components/TripReportCard.tsx [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            <RichTextRenderer data={report.content} />
```
[the remark this answers is not in the document]
Comment (neryadom) [reply src/app/(frontend)/trip-reports/_components/TripReportCard.tsx:57]: Yes, noticed that yesterday as well, links in the rich text cause issues... Maybe we could hack around way around it by making each card a div instead of a Link element? Thus avoiding the nested link problem. But yes, as you say, going straight for text makes 100% more sense at this stage.
